be/dtm0_log.h

```diff
@@ -0,0 +1,236 @@
+ *
+ *  During recovery operation of any of the participant, rest of the participant
+ *  and originator will iterate over the logged journal and extract the state
+ *  of each transation for participant under recovery from logged information
+ *  and will send redo request for those transaction which  are not
+ *  M0_DTML_STATE_PERSISTENT on participant being recovered.
+ *
+ *  Upon receiving redo request participant under recovery will log the state
+ *  of same transaction of remote participant in persistent store.
+ *
+ * @section Usecases
+ *
+ * 1. When transaction successfully executed on participant, DTM0 log will be
+ *    added by participant to persistent store and state of transaction will be
+ *    M0_DTML_STATE_EXECUTED, if log is create first time then state for same
+ *    transaction for remote participant will be logged as M0_DTML_STATE_UNKNOWN
+ *    later upon getting persistent notice from remote participant appropriate
+ *    state for particular participant will be updated.
+ *
+ * {
+ *    struct log*   l;
+ *    struct be_tx* tx;
+ *    struct be_tx_credit cred;
+ *
+ *    log_tx_credit(cred)
+ *    tx_open(tx, cred);
+ *
+ *    if (!log_already_present) {
+ *	l = log_create(tx);
+ *      txr = m0_be_dtm0_log_find(l, tx_id);
+ *    } else {
+ *      txr = m0_be_dtm0_log_find(l, tx_id);
+ *    }
+ *    ...
+ *    log_update(l, tx, M0_DTML_STATE_EXECUTED, (struct txr*){ ... });
+ *    tx_close(tx);
+ * }
+ *
+ * 2. When transaction become persistent on particular participant, the state
+ *    of the transaction for this participant will be updated as
+ *    M0_DTML_STATE_PERSISTENT.
+ *
+ * {
+ *    struct log*   l;
+ *    struct be_tx* tx;
+ *    struct be_tx_credit cred;
+ *
+ *    log_tx_credit(cred)
+ *    tx_open(tx, cred);
+ *
+ *    txr = m0_be_dtm0_log_find(l, tx_id);
+ *
+ *    ...
+ *    log_update(l, tx, M0_DTML_STATE_PERSISTENT, (struct txr*){ ... });
+ *    tx_close(tx);
+ * }
+ *
+ * 3. When transaction become persistent on remote participant, it will send the
+ *    persistent notice to rest of the participant and originator stating that
+ *    transaction is persistent on it's store. Upon receiving this message each
+ *    of participant and originators will update the state of transaction for
+ *    remote participant as M0_DTML_STATE_PERSISTENT.
+ *
+ *    If persistent message from remote participant arrives before receiving
+ *    participant or originator has logged the same transaction on it's
+ *    persistent/volatile store then with this event log will be created and
+ *    remote participant state and txr id will be logged but txr payload will
+ *    be invalid as persistent message from remote participant does not carry
+ *    payload.
+ *
+ * {
+ *    struct log*   l;
+ *    struct be_tx* tx;
+ *    struct be_tx_credit cred;
+ *
+ *    log_tx_credit(cred)
+ *    tx_open(tx, cred);
+ *
+ *    if (!log_already_present) {
+ *	l = log_create(tx);
```

Comment:
Q: Shouldn't we create DTM0 log as a part of mkfs or during the startup procedure?

Reply:
Addressed at 4e0d3688b2cc72748d9c1ed8df1ddeb8a2319fba